.jenkins/infrastructure/build_azure_managed_images.Jenkinsfile

```diff
@@ -84,36 +86,42 @@ def buildLinuxManagedImage(String os_type, String version) {
 
 def buildWindowsManagedImage(String os_series, String img_name_suffix, String launch_configuration) {
     node(params.AGENTS_LABEL) {
-        stage(img_name_suffix) {
-            timeout(GLOBAL_TIMEOUT_MINUTES) {
+        withCredentials([usernamePassword(credentialsId: JENKINS_USER_CREDS_ID,
+                                    usernameVariable: "JENKINS_USER_NAME",
+                                    passwordVariable: "JENKINS_USER_PASSWORD")]) {
+            def az_login_script = """
+                az login --service-principal -u \$SERVICE_PRINCIPAL_ID -p \$SERVICE_PRINCIPAL_PASSWORD --tenant \$TENANT_ID
+                az account set -s \$SUBSCRIPTION_ID
+            """
+            def managed_image_name_id = get_image_id()
+            def gallery_image_version = get_image_version()
+            def vm_rg_name = "build-${managed_image_name_id}-${img_name_suffix}-${BUILD_NUMBER}"
+            def vm_name = "${os_series}-vm"
+            def jenkins_rg_name = params.JENKINS_RESOURCE_GROUP
+            def jenkins_vnet_name = params.JENKINS_VNET_NAME
+            def jenkins_subnet_name = params.JENKINS_SUBNET_NAME
+            def azure_image_id = AZURE_IMAGES_MAP[os_series]["image"]
+
+            stage("${img_name_suffix}-prepare") {
+                def az_rg_create_script = """
+                    ${az_login_script}
+                    az group create --name ${vm_rg_name} --location ${REGION}
+                """
                 cleanWs()
-                checkout scm
-                def managed_image_name_id = get_image_id()
-                def gallery_image_version = get_image_version()
-                def vm_rg_name = "build-${managed_image_name_id}-${img_name_suffix}-${BUILD_NUMBER}"
-                def vm_name = "${os_series}-vm"
-                def jenkins_rg_name = params.JENKINS_RESOURCE_GROUP
-                def jenkins_vnet_name = params.JENKINS_VNET_NAME
-                def jenkins_subnet_name = params.JENKINS_SUBNET_NAME
-                def azure_image_id = AZURE_IMAGES_MAP[os_series]["image"]
-                withCredentials([usernamePassword(credentialsId: JENKINS_USER_CREDS_ID,
-                                                  usernameVariable: "JENKINS_USER_NAME",
-                                                  passwordVariable: "JENKINS_USER_PASSWORD")]) {
-                    def az_login_script = """
-                        az login --service-principal -u \$SERVICE_PRINCIPAL_ID -p \$SERVICE_PRINCIPAL_PASSWORD --tenant \$TENANT_ID
-                        az account set -s \$SUBSCRIPTION_ID
-                    """
-                    def az_build_managed_img_script = """
-                        source ${WORKSPACE}/.jenkins/infrastructure/provision/utils.sh
+                oe.azureEnvironment(az_rg_create_script, params.OE_DEPLOY_IMAGE)
+            }
 
+            try {
+                stage("${img_name_suffix}-provisioning") {
+                    def provision_script = """
```

Comment:
Is it possible/worth it having these types of multi-line commands in a script instead? Having it in a script seems easier to maintain, unless that type of workflow doesn't work, or is too much trouble.

Reply:
Interesting idea to think about. Thanks @rs-- !

As we use a global library function to run commands/scripts inside an Azure environment and the command/script needs to be passed as a parameter each time, the consequences of splitting into individual stages is that the big script that we started with also has to be broke up. The advantage is that each part of the script can be ran separately and retried as needed.

It is then possible to move these functions into separate scripts by treating them like libraries similar to this: https://github.com/openenclave/openenclave-ci/blob/master/src/jenkins/common/Openenclave.groovy
The downside to that is that adds some overhead as the libraries need to then be loaded in with the Jenkins master and/or in the Jenkinsfile, which then becomes a trade off between which files/configuration do we want to maintain. I can't say for sure which one would be easier to maintain, but currently I think having the scripts in this single Jenkinsfile at least makes it easier to find and read. Would you agree with that?

Reply:
I'd say that if the functionality was generic enough like provisioning a VM from scratch, it would be a benefit to have it split out since we could use it in other places (like when we have to manually create a VM). But if we're only doing this here, we can just leave it, and split it out in the future when the functionality is needed elsewhere.

Reply:
A good plan. Thanks!